File: lib/models/message.dart
class Message {
  final String id;
  final String senderId;
  final String senderName;
  final String senderAvatar;
  final String content;
  final DateTime timestamp;
  final String? channelId;
  final String? dmPartnerId;

  const Message({
    required this.id,
    required this.senderId,
    required this.senderName,
    required this.senderAvatar,
    required this.content,
    required this.timestamp,
    this.channelId,
    this.dmPartnerId,
  });

  factory Message.fromJson(Map<String, dynamic> json) {
    return Message(
      id: json['id'] ?? json['_id'] ?? '',
      senderId: json['senderId'] ?? '',
      senderName: json['senderName'] ?? '',
      senderAvatar: json['senderAvatar'] ?? '',
      content: json['content'] ?? '',
      timestamp: DateTime.tryParse(json['timestamp'] ?? '') ?? DateTime.now(),
      channelId: json['channelId'],
      dmPartnerId: json['dmPartnerId'],
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'senderId': senderId,
    'senderName': senderName,
    'senderAvatar': senderAvatar,
    'content': content,
    'timestamp': timestamp.toIso8601String(),
    'channelId': channelId,
    'dmPartnerId': dmPartnerId,
  };
}

File: backend/models/Message.js
const mongoose = require('mongoose');

const messageSchema = new mongoose.Schema({
  senderId:    { type: String, required: true },
  senderName:  { type: String, required: true },
  senderAvatar:{ type: String, default: '' },
  content:     { type: String, required: true },
  channelId:   { type: String, default: null },
  dmPartnerId: { type: String, default: null },
  // For DMs we store both participant IDs for querying
  dmParticipants: [{ type: String }],
}, { timestamps: true });

messageSchema.methods.toJSON = function() {
  const obj = this.toObject();
  obj.id = obj._id.toString();
  obj.timestamp = obj.createdAt;
  delete obj._id;
  delete obj.__v;
  delete obj.updatedAt;
  return obj;
};

// Index for efficient queries
messageSchema.index({ channelId: 1, createdAt: 1 });
messageSchema.index({ dmParticipants: 1, createdAt: 1 });

module.exports = mongoose.model('Message', messageSchema);
